### PR TITLE
Add failIfNoTests attribute to @Suite

### PR DESCRIPTION
**Pull request by mpkorstanje**
## Overview

Adds new `failIfNoTests` attribute to `@Suite`. This will fail the suite if no tests were discovered, resolving part of #2819.
    
This mechanism is implemented by having the `SummaryGeneratingListener` listen in on the test execution while also delegating any execution events to the engine execution listener of the parent engine. If the listener did not hear any test executions the suite will mark its own execution as a failure. This is identical to the console launcher.
    
To prevent pruning the `SuiteTestDescriptor` from being pruned prematurely and not being executed at all, the suite  describes
itself a as test container that may register test. In practice the container will never register any tests.
    
Closes: #2823

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/CompositeEngineExecutionListener.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.reporting.ReportEntry;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+class CompositeEngineExecutionListener implements EngineExecutionListener {
```
Comment (mpkorstanje) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/CompositeEngineExecutionListener.java:26]: TODO: Missing test case.

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineExecutionOrchestrator.java [hunk 2]
```diff
@@ -45,13 +46,20 @@ public EngineExecutionOrchestrator() {
 	}
 
 	void execute(InternalTestPlan internalTestPlan, TestExecutionListener... listeners) {
+		execute(internalTestPlan, EngineExecutionListener.NOOP, listeners);
+	}
+
+	public void execute(InternalTestPlan internalTestPlan, EngineExecutionListener parentEngineExecutionListener, TestExecutionListener... listeners) {
```
Comment (mpkorstanje) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineExecutionOrchestrator.java:52]: TODO: Consider exposing `InternalTestPlan` vs exposing `ListenerRegistry`.

**Comment by sormuras** [comment]
Restarted the failing "Cross Version" check twice, it doesn't seem to pick up the fixed JDK-URI mapping file. The fixed mapping file contains the entry needed by check:

https://github.com/oracle-actions/setup-java/blob/0fe26a3d562e2aae686cfd37ebb48cc6602a233a/jdk.java.net-uri.properties#L52

**Comment by sormuras** [comment]
image

3rd restart did it.

Caches?

**Comment by mpkorstanje** [comment]
> Caches?

It would appear to be so. Though I was thinking the pipeline might be immutable.

**Review by sormuras — APPROVED** [review]
LGTM

**Comment by sormuras** [comment]
Thanks for this PR, @mpkorstanje

**Pull request merged**